pymc3/variational/approximations.py

```diff
@@ -322,47 +338,46 @@ def randidx(self, size=None):
                          high=pm.floatX(self.histogram.shape[0]) - pm.floatX(1e-16))
                 .astype('int32'))
 
-    def random_global(self, size=None, no_rand=False):
-        theano_condition_is_here = isinstance(no_rand, tt.Variable)
-        if theano_condition_is_here:
-            return tt.switch(no_rand,
-                             self.mean,
-                             self.histogram[self.randidx(size)])
-        else:
-            if no_rand:
-                return self.mean
+    def _initial_part_matrix(self, part, size, deterministic):
+        if part == 'local':
+            return super(Empirical, self)._initial_part_matrix(
+                part, size, deterministic
+            )
+        elif part == 'global':
+            theano_condition_is_here = isinstance(deterministic, tt.Variable)
+            if theano_condition_is_here:
+                return tt.switch(
+                    deterministic,
+                    tt.repeat(
+                        self.mean.dimshuffle('x', 0),
+                        size if size is not None else 1, -1),
+                    self.histogram[self.randidx(size)])
             else:
-                return self.histogram[self.randidx(size)]
+                if deterministic:
+                    return tt.repeat(
+                        self.mean.dimshuffle('x', 0),
+                        size if size is not None else 1, -1)
+                else:
+                    return self.histogram[self.randidx(size)]
 
     @property
-    def histogram(self):
-        """Shortcut to flattened Trace
-        """
-        return self.shared_params
+    def symbolic_random_global_matrix(self):
+        return self.symbolic_initial_global_matrix
 
     @property
-    @memoize
-    def histogram_logp(self):
-        """Symbolic logp for every point in trace
+    def histogram(self):
```

Comment:
As I remember we decided that Empirical is about what is a class about and Histogram is about how it stores samples. I see no bad in such property